src/test/parallel.py

```diff
@@ -0,0 +1,864 @@
+    milliseconds, or None if the test failed or if there's no record for it."""
+    return self.__times.get((binary, testname), None)
+
+  def record_test_time(self, binary, testname, runtime_ms):
+    """Record that the given test ran in the specified number of
+    milliseconds. If the test failed, runtime_ms should be None."""
+    with self.__lock:
+      self.__times[(binary, testname)] = runtime_ms
+
+  def write_to_file(self, save_file):
+    "Write all the times to file."
+    try:
+      with TestTimes.LockedFile(save_file, 'a+b') as fd:
+        times = TestTimes.__read_test_times_file(fd)
+
+        if times is None:
+          times = self.__times
+        else:
+          times.update(self.__times)
+
+        # We erase data from file while still holding a lock to it. This
+        # way reading old test times and appending new ones are atomic
+        # for external viewer.
+        fd.seek(0)
+        fd.truncate()
+        with gzip.GzipFile(fileobj=fd, mode='wb') as gzf:
+          cPickle.dump(times, gzf, PICKLE_HIGHEST_PROTOCOL)
+    except IOError:
+      pass  # ignore errors---saving the times isn't that important
+
+  @staticmethod
+  def __read_test_times_file(fd):
+    try:
+      with gzip.GzipFile(fileobj=fd, mode='rb') as gzf:
+        times = cPickle.load(gzf)
+    except Exception:
+      # File doesn't exist, isn't readable, is malformed---whatever.
+      # Just ignore it.
+      return None
+    else:
+      return times
+
+
+def find_tests(test_list, binaries, additional_args, options, times):
+  test_count = 0
+  tasks = []
+  for test_binary in binaries:
+    command = [test_binary]
+    command += additional_args
+
+    for test_name in test_list:
+      last_execution_time = times.get_test_time(test_binary, test_name)
+      if options.failed and last_execution_time is not None:
+        continue
+
+      test_command = command + ['--run_test=' + test_name]
+      if (test_count - options.shard_index) % options.shard_count == 0:
+        for execution_number in range(options.repeat):
+          tasks.append(Task(test_binary, test_name, test_command,
+                            execution_number + 1, last_execution_time,
+                            options.output_dir))
+
+      test_count += 1
+
+  # Sort the tasks to run the slowest tests first, so that faster ones can be
+  # finished in parallel.
+  return sorted(tasks, reverse=True)
+
+
+def execute_tasks(tasks, pool_size, task_manager,
+                  timeout, serialize_test_cases):
+  class WorkerFn(object):
+    def __init__(self, tasks, running_groups):
+      self.tasks = tasks
+      self.running_groups = running_groups
+      self.task_lock = threading.Lock()
+
+    def __call__(self):
+      while True:
+        with self.task_lock:
+          for task_id in range(len(self.tasks)):
+            task = self.tasks[task_id]
+
+            if self.running_groups is not None:
+              test_group = task.test_name.split('.')[0]
```

Comment:
Should this be `split('/')[0]` ?

Right now `task.test_name` is a line from `src/test/test_list.txt` like 
```
bloom_tests/rolling_bloom
```
So the split will always return that full path, since there is no `.` - making `option.serialize_test_cases` just run everything in parallel anyway.